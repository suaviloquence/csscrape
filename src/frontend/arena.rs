use std::{fmt::Debug, marker::PhantomData, num::NonZeroUsize};

#[derive(Debug, Clone)]
pub struct ArenaInner<T> {
    inner: Vec<T>,
    /// we use 1-indexed refs for null-pointer optimization
    next_ref: NonZeroUsize,
}

#[derive(Debug)]
pub struct Arena<T>(ArenaInner<T>);

#[derive(Debug)]
pub struct Ref<'a, T>(NonZeroUsize, PhantomData<&'a T>);

/// explicit impl because we want it even when `T` is not Clone
impl<'a, T> Clone for Ref<'a, T> {
    fn clone(&self) -> Self {
        Self(self.0, PhantomData)
    }
}

impl<'a, T> Copy for Ref<'a, T> {}

impl<T> Arena<T> {
    #[inline]
    #[must_use]
    pub const fn new() -> Self {
        Self(ArenaInner {
            inner: Vec::new(),
            // safety: 1 is nonzero
            next_ref: unsafe { NonZeroUsize::new_unchecked(1) },
        })
    }

    #[inline]
    #[must_use]
    pub fn get<'a, 's>(&'s self, r: Ref<'a, T>) -> &'s T
    where
        'a: 's,
    {
        &self.0.inner[r.0.get() - 1]
    }

    pub fn insert<'a, 's>(&'s mut self, item: T) -> Ref<'a, T>
    where
        'a: 's,
    {
        let r = Ref(self.0.next_ref, PhantomData);
        self.0.next_ref = self
            .0
            .next_ref
            .checked_add(1)
            .expect("maximum elements in arena.");
        self.0.inner.push(item);
        r
    }
}

#[cfg(test)]
mod tests {
    use super::Arena;

    #[test]
    fn test_compiles() {
        let mut arena = Arena::<u8>::new();
        let r1 = arena.insert(1);
        let r2 = arena.insert(2);
        assert_eq!(*arena.get(r1), 1);
        assert_eq!(*arena.get(r2), 2);
    }

    #[test]
    fn test_no_copy() {
        #[derive(Debug, PartialEq, Eq)]
        struct NoCopy(u8);

        let mut arena = Arena::new();
        let r1 = arena.insert(NoCopy(1));
        let r2 = arena.insert(NoCopy(2));

        assert_eq!(&NoCopy(1), arena.get(r1));
        assert_eq!(&NoCopy(2), arena.get(r2));
    }
}
