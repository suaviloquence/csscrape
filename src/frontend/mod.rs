mod arena;
pub mod ast;
mod parser;
mod scanner;

pub use parser::{ParseError, Parser};
