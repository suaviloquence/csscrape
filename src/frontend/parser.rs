use core::fmt;
use std::borrow::Cow;

use super::{
    arena::Arena,
    ast::{
        ArgList, Ast, AstRef, Element, ElementList, ElementStatementList, FilterList, Leaf,
        Selector, SelectorCombinator, SelectorList, SelectorOpts, Statement,
    },
    scanner::{Lexeme, Scanner, Token},
};

#[derive(Debug)]
pub struct Parser<'a> {
    scanner: Scanner<'a>,
    arena: Arena<'a, Ast<'a>>,
}

#[derive(Debug, Clone)]
#[non_exhaustive]
pub enum ParseError<'a> {
    UnexpectedToken {
        expected: Vec<Token>,
        got: Lexeme<'a>,
    },
}

impl<'a> fmt::Display for ParseError<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::UnexpectedToken { expected, got } => {
                write!(f, "Expected one of {expected:?}, got {got:?}")
            }
        }
    }
}

type Result<'a, T> = std::result::Result<T, ParseError<'a>>;

impl<'a> Parser<'a> {
    #[must_use]
    pub const fn new(input: &'a str) -> Self {
        Self {
            scanner: Scanner::new(input),
            arena: Arena::new(),
        }
    }

    pub fn parse(
        mut self,
    ) -> Result<'a, (Arena<'a, Ast<'a>>, Option<AstRef<'a, ElementList<'a>>>)> {
        let r = self.parse_element_list()?;
        self.try_eat(Token::Eof)?;
        Ok((self.arena, r))
    }

    fn parse_element_list(&mut self) -> Result<'a, Option<AstRef<'a, ElementList<'a>>>> {
        let lx = self.scanner.peek_non_whitespace();

        match lx.token {
            Token::Eof => Ok(None),
            Token::Id | Token::Dot | Token::Hash | Token::Star => {
                let element = self.parse_element()?;
                let next = self.parse_element_list()?;
                let r = self.arena.insert_variant(ElementList::new(element, next));
                Ok(Some(r))
            }
            _ => Err(ParseError::UnexpectedToken {
                expected: vec![Token::Eof, Token::Id, Token::Dot, Token::Hash, Token::Star],
                got: lx,
            }),
        }
    }

    #[inline]
    fn try_eat(&mut self, tk: Token) -> Result<'a, Lexeme<'a>> {
        let lx = self.scanner.peek_non_whitespace();
        self.scanner.eat_token();

        if lx.token == tk {
            Ok(lx)
        } else {
            Err(ParseError::UnexpectedToken {
                expected: vec![tk],
                got: lx,
            })
        }
    }

    fn parse_element(&mut self) -> Result<'a, Element<'a>> {
        let selector_head = self.parse_selector()?;
        let selectors = self.parse_selector_list()?;
        let lx = self.scanner.peek_non_whitespace();

        let ops = match lx.token {
            Token::Question => {
                self.scanner.eat_token();
                SelectorOpts::Optional
            }
            Token::Collection => {
                self.scanner.eat_token();
                SelectorOpts::Collection
            }
            _ => SelectorOpts::One,
        };

        self.try_eat(Token::BraceOpen)?;

        let statements = self.parse_element_statement_list()?;

        self.try_eat(Token::BraceClose)?;

        Ok(Element {
            selector_head,
            ops,
            selectors,
            statements,
        })
    }

    fn parse_selector_list(&mut self) -> Result<'a, Option<AstRef<'a, SelectorList<'a>>>> {
        let mut lx = self.scanner.peek_token();
        if lx.token == Token::Whitespace {
            self.scanner.eat_token();
            let next_lx = self.scanner.peek_non_whitespace();
            // if the next lexeme after the whitespace doesn't signify a selector,
            // the whitespace is not significant.
            match next_lx.token {
                Token::Id | Token::Hash | Token::Dot | Token::Star => (),
                _ => lx = next_lx,
            };
        }

        let sel = match lx.token {
            Token::BraceOpen => return Ok(None),
            // invariant: peek_next_whitespace is one of Id | Hash | Dot | Star
            // whitespace is eaten in the above block.
            Token::Whitespace => SelectorCombinator::Descendent(self.parse_selector()?),
            Token::Greater => {
                self.scanner.eat_token();
                SelectorCombinator::Child(self.parse_selector()?)
            }
            Token::Plus => {
                self.scanner.eat_token();
                SelectorCombinator::NextSibling(self.parse_selector()?)
            }
            Token::Tilde => {
                self.scanner.eat_token();
                SelectorCombinator::SubsequentSibling(self.parse_selector()?)
            }
            Token::Hash | Token::Dot | Token::Id | Token::Star => {
                SelectorCombinator::And(self.parse_selector()?)
            }
            _ => {
                return Err(ParseError::UnexpectedToken {
                    expected: vec![
                        Token::Whitespace,
                        Token::Greater,
                        Token::Plus,
                        Token::Tilde,
                        Token::Hash,
                        Token::Dot,
                        Token::Id,
                        Token::Star,
                    ],
                    got: lx,
                })
            }
        };

        let itm = SelectorList::new(sel, self.parse_selector_list()?);

        Ok(Some(self.arena.insert_variant(itm)))
    }

    fn parse_selector(&mut self) -> Result<'a, Selector<'a>> {
        let lx = self.scanner.peek_non_whitespace();
        match lx.token {
            Token::Dot => {
                self.scanner.eat_token();
                self.try_eat(Token::Id).map(|lx| Selector::Class(lx.value))
            }
            Token::Hash => {
                self.scanner.eat_token();
                self.try_eat(Token::Id).map(|lx| Selector::Id(lx.value))
            }
            Token::Id => {
                self.scanner.eat_token();
                Ok(Selector::Tag(lx.value))
            }
            Token::Star => {
                self.scanner.eat_token();
                Ok(Selector::Any)
            }
            _ => Err(ParseError::UnexpectedToken {
                expected: vec![Token::Dot, Token::Hash, Token::Id, Token::Star],
                got: lx,
            }),
        }
    }

    fn parse_element_statement_list(
        &mut self,
    ) -> Result<'a, Option<AstRef<'a, ElementStatementList<'a>>>> {
        let lx = self.scanner.peek_non_whitespace();
        let item = match lx.token {
            Token::At => Err(self.parse_statement()?),
            Token::Id | Token::Dot | Token::Hash | Token::Star => Ok(self.parse_element()?),
            Token::BraceClose | Token::Eof => return Ok(None),
            _ => {
                return Err(ParseError::UnexpectedToken {
                    expected: vec![
                        Token::At,
                        Token::Id,
                        Token::Dot,
                        Token::Hash,
                        Token::Star,
                        Token::BraceClose,
                        Token::Eof,
                    ],
                    got: lx,
                })
            }
        };

        let next = self.parse_element_statement_list()?;
        Ok(Some(
            self.arena
                .insert_variant(ElementStatementList::new(item, next)),
        ))
    }

    fn parse_statement(&mut self) -> Result<'a, Statement<'a>> {
        self.try_eat(Token::At)?;
        let id = self.try_eat(Token::Id)?.value;
        self.try_eat(Token::Colon)?;
        let value = self.try_eat(Token::Id)?.value;
        let filters = self.parse_filter_list()?;
        Ok(Statement { id, value, filters })
    }

    fn parse_filter_list(&mut self) -> Result<'a, Option<AstRef<'a, FilterList<'a>>>> {
        let lx = self.scanner.peek_non_whitespace();
        if lx.token == Token::Pipe {
            self.scanner.eat_token();
            let id = self.try_eat(Token::Id)?.value;
            self.try_eat(Token::ParenOpen)?;
            let args = self.parse_arg_list()?;
            self.try_eat(Token::ParenClose)?;
            let next = self.parse_filter_list()?;
            let r = self.arena.insert_variant(FilterList::new(id, args, next));
            Ok(Some(r))
        } else {
            Ok(None)
        }
    }

    fn parse_arg_list(&mut self) -> Result<'a, Option<AstRef<'a, ArgList<'a>>>> {
        let lx = self.scanner.peek_non_whitespace();
        match lx.token {
            Token::ParenClose => Ok(None),
            Token::Id => {
                let id = lx.value;
                self.scanner.eat_token();
                self.try_eat(Token::Colon)?;
                let lx = self.scanner.peek_non_whitespace();
                let value = match lx.token {
                    Token::Id => Leaf::Id(lx.value),
                    Token::Float => Leaf::Float(lx.value.parse().expect("float should be valid")),
                    Token::Int => Leaf::Int(lx.value.parse().expect("int should be valid")),
                    Token::String => Leaf::String(parse_string_literal(lx.value)),
                    _ => {
                        return Err(ParseError::UnexpectedToken {
                            expected: vec![Token::Id, Token::Float, Token::Int, Token::String],
                            got: lx,
                        })
                    }
                };
                self.scanner.eat_token();
                let next = match self.scanner.peek_non_whitespace().token {
                    Token::Comma => {
                        self.scanner.eat_token();
                        self.parse_arg_list()?
                    }
                    _ => None,
                };

                let r = self.arena.insert_variant(ArgList::new(id, value, next));
                Ok(Some(r))
            }
            _ => Err(ParseError::UnexpectedToken {
                expected: vec![Token::ParenClose, Token::Id],
                got: lx,
            }),
        }
    }
}

fn parse_string_literal(s: &str) -> Cow<'_, str> {
    debug_assert!(s.len() >= 2 && &s[0..1] == "\"" && &s[s.len() - 1..] == "\"");
    let mut prune = vec![];
    let s = &s[1..s.len() - 1];

    let mut escape_next = false;
    for (i, s) in s.char_indices() {
        if escape_next {
            escape_next = false;
        } else if s == '\\' {
            escape_next = true;
            prune.push(i);
        }
    }

    if prune.is_empty() {
        Cow::Borrowed(s)
    } else {
        Cow::Owned(
            s.char_indices()
                .filter_map(|(i, x)| prune.contains(&i).then_some(x))
                .collect(),
        )
    }
}

#[cfg(test)]
mod tests {
    use std::borrow::Cow;

    use super::Parser;
    use crate::frontend::ast::*;

    fn fmt_selector<'a>(head: &Selector<'a>, list: &[&SelectorList<'a>]) -> String {
        use std::fmt::Write as _;
        let mut out = String::new();
        write!(&mut out, "{head}").expect("fmt error");
        for node in list {
            match &node.sel {
                SelectorCombinator::And(s) => write!(&mut out, "{s}"),
                SelectorCombinator::Child(s) => write!(&mut out, " > {s}"),
                SelectorCombinator::Descendent(s) => write!(&mut out, " {s}"),
                SelectorCombinator::NextSibling(s) => write!(&mut out, " + {s}"),
                SelectorCombinator::SubsequentSibling(s) => write!(&mut out, " ~ {s}"),
            }
            .expect("fmt error");
        }

        out
    }

    #[test]
    fn test_elements() {
        let string = r#"h1 {
            @x: me | cat(i: "x", ) | meow()
            h2#x > .cat  {
            }
            }"#;
        let parser = Parser::new(&string);
        let (arena, r) = parser.parse().expect("parsing failed");

        let elements = arena.flatten(r);
        let element = &elements[0].value;

        assert_eq!(
            fmt_selector(&element.selector_head, &arena.flatten(element.selectors)),
            "h1"
        );

        assert_eq!(element.ops, SelectorOpts::One);
        let statements = arena.flatten(element.statements);
        let stmt = statements[0]
            .value
            .as_ref()
            .expect_err("should be a statement");
        assert!(
            matches!(
                stmt,
                Statement {
                    id: "x",
                    value: "me",
                    ..
                }
            ),
            "found {stmt:?}",
        );

        let filters = arena.flatten(stmt.filters);
        assert!(
            matches!(
                &filters[..],
                [FilterList { id: "cat", .. }, FilterList { id: "meow", .. }]
            ),
            "found {filters:?}"
        );

        let filter = filters[0];
        let args = arena.flatten(filter.args);
        assert!(
            matches!(
                &args[..],
                [ArgList {
                    id: "i",
                    value: Leaf::String(Cow::Borrowed("x")),
                    ..
                }]
            ),
            "found {:?}",
            &args[..]
        );

        let element = statements[1].value.as_ref().expect("should b an element");
        assert!(element.statements.is_none());
        assert_eq!(
            fmt_selector(&element.selector_head, &arena.flatten(element.selectors)),
            "h2#x > .cat"
        );
    }
}
