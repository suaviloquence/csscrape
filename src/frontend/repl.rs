#[derive(Debug, Clone)]
pub struct Repl {}
