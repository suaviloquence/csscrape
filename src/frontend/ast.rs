use core::{fmt, marker::PhantomData};
use std::borrow::Cow;

use super::arena::{Arena, Ref};

pub type AstArena<'a> = Arena<Ast<'a>>;

#[derive(Debug)]
pub struct AstRef<'a, T>(Ref<'a, Ast<'a>>, PhantomData<&'a T>);

pub trait AstType<'a>: Sized {
    #[must_use]
    fn unwrap_ref<'b>(node: &'b Ast<'a>) -> &'b Self;

    #[must_use]
    fn wrap(self) -> Ast<'a>;
}

pub trait AstArenaFlatten<'a>: AstType<'a> {
    fn flatten<'s: 'o, 'o>(&'s self, arena: &'s AstArena<'a>, out: &mut Vec<&'o Self>);
}

/// explicit impl because we want it even when `T` is not Clone
impl<'a, T> Clone for AstRef<'a, T> {
    fn clone(&self) -> Self {
        *self
    }
}

impl<'a, T> Copy for AstRef<'a, T> {}

/// specialization for typed references
impl<'a> AstArena<'a> {
    #[inline]
    #[must_use]
    pub fn get_variant<'s, T>(&'s self, r: AstRef<'a, T>) -> &'s T
    where
        'a: 's,
        T: AstType<'a>,
    {
        T::unwrap_ref(self.get(r.0))
    }

    pub fn insert_variant<'s, T: AstType<'a>>(&'s mut self, variant: T) -> AstRef<'a, T>
    where
        'a: 's,
    {
        let r = self.insert(variant.wrap());
        AstRef(r, PhantomData)
    }

    #[must_use]
    pub fn flatten<T: AstArenaFlatten<'a>>(&self, r: Option<AstRef<'a, T>>) -> Vec<&T> {
        let mut out = Vec::new();
        if let Some(r) = r {
            self.get_variant(r).flatten(self, &mut out);
        }
        out
    }
}

#[derive(Debug, Clone)]
pub enum Selector<'a> {
    Any,
    Tag(&'a str),
    Class(&'a str),
    Id(&'a str),
}

impl fmt::Display for Selector<'_> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Any => write!(f, "*"),
            Self::Tag(id) => write!(f, "{id}"),
            Self::Class(id) => write!(f, ".{id}"),
            Self::Id(id) => write!(f, "#{id}"),
        }
    }
}

#[derive(Debug, Clone)]
pub enum SelectorCombinator<'a> {
    NextSibling(Selector<'a>),
    Child(Selector<'a>),
    SubsequentSibling(Selector<'a>),
    Descendent(Selector<'a>),
    And(Selector<'a>),
}

impl fmt::Display for SelectorCombinator<'_> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::NextSibling(s) => write!(f, " + {s}"),
            Self::Child(s) => write!(f, " > {s}"),
            Self::SubsequentSibling(s) => write!(f, " ~ {s}"),
            Self::Descendent(s) => write!(f, " {s}"),
            Self::And(s) => write!(f, "{s}"),
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub enum Qualifier {
    #[default]
    One,
    Optional,
    Collection,
}

macro_rules! ast_enum {
        {
            // macro limitations can only have nonrepeating metavariable  at this scope
            // see https://github.com/rust-lang/rust/issues/96184
            #![$allmeta: meta]
            $(#[$astmeta: meta])*
            pub enum Ast<'a> {
                $(
                    $(#[$indivmeta: meta])*
                    $(@flatten[$(.$preorder: ident, )* self $(, .$postorder: ident)*])?
                    $variant: ident {
                        $(
                            $(#[$membermeta: meta])*
                            $member: ident: $ty: ty,
                        )*
                    },
                )*
            }
        } => {
            $(
                #[$allmeta]
                #[non_exhaustive]
                $(#[$indivmeta])*
                pub struct $variant<'a> {
                    $(
                        $(#[$membermeta])*
                        pub $member: $ty,
                    )*
                    lt: PhantomData<&'a ()>,
                }

                impl<'a> $variant<'a> {
                    #[must_use]
                    #[inline]
                    pub const fn new($($member: $ty, )*) -> Self {
                        Self { lt: PhantomData, $($member, )* }
                    }
                }

                impl<'a> AstType<'a> for $variant<'a> {
                    #[inline]
                    #[must_use]
                    fn unwrap_ref<'b>(node: &'b Ast<'a>) -> &'b Self {
                        let Ast::$variant(x) = node else { unreachable!(concat!("expected a `Ast::", stringify!($variant), "` variant."))};
                        x
                    }

                    #[inline]
                    #[must_use]
                    fn wrap(self) -> Ast<'a> {
                        Ast::$variant(self)
                    }
                }

                $(
                    impl<'a> AstArenaFlatten<'a> for $variant<'a> {
                        fn flatten<'s: 'o, 'o>(&'s self, arena: &'s AstArena<'a>, out: &mut Vec<&'o Self>) {
                            $(
                                if let Some(pre) = self.$preorder {
                                    arena.get_variant(pre).flatten(arena, out);
                                }
                            )*
                            out.push(self);
                            $(
                                if let Some(post) = self.$postorder {
                                    arena.get_variant(post).flatten(arena, out);
                                }
                            )*
                        }
                    }
                )?
            )*

            #[$allmeta]
            $(#[$astmeta])*
            pub enum Ast<'a> {
                $($variant($variant<'a>),)*
            }
        };
    }

#[derive(Debug, Clone)]
pub enum Leaf<'a> {
    Var(&'a str),
    Int(i64),
    Float(f64),
    String(Cow<'a, str>),
}

#[derive(Debug, Clone)]
pub enum RValue<'a> {
    Leaf(Leaf<'a>),
    Element(Element<'a>),
}

#[derive(Debug, Clone)]
pub struct Inline<'a> {
    pub value: Leaf<'a>,
    pub filters: Option<AstRef<'a, FilterList<'a>>>,
}

#[derive(Debug, Clone)]
pub enum Value<'a> {
    Leaf(Leaf<'a>),
    Inline(Inline<'a>),
}

#[derive(Debug, Clone)]
pub struct Statement<'a> {
    pub id: &'a str,
    pub value: RValue<'a>,
    pub filters: Option<AstRef<'a, FilterList<'a>>>,
}

#[derive(Debug, Clone)]
pub struct Element<'a> {
    pub url: Option<Inline<'a>>,
    pub selector_head: Selector<'a>,
    pub selectors: Option<AstRef<'a, SelectorList<'a>>>,
    pub qualifier: Qualifier,
    pub statements: Option<AstRef<'a, StatementList<'a>>>,
}

#[derive(Debug, Clone)]
pub enum Filter<'a> {
    Call(FilterCall<'a>),
    Select(FilterSelect<'a>),
}

ast_enum! {
#![derive(Debug, Clone)]
pub enum Ast<'a> {
    @flatten[self, .next]
    SelectorList {
        sel: SelectorCombinator<'a>,
        next: Option<AstRef<'a, SelectorList<'a>>>,
    },
    @flatten[self, .next]
    ArgList {
        id: &'a str,
        value: Leaf<'a>,
        next: Option<AstRef<'a, ArgList<'a>>>,
    },
    FilterCall {
        id: &'a str,
        args: Option<AstRef<'a, ArgList<'a>>>,
    },
    FilterSelect {
        name: &'a str,
        value: Value<'a>,
    },
    @flatten[self, .next]
    FilterList {
        filter: Filter<'a>,
        qualifier: Qualifier,
        next: Option<AstRef<'a, FilterList<'a>>>,
    },
    @flatten[self, .next]
    StatementList {
        value: Statement<'a>,
        next: Option<AstRef<'a, StatementList<'a>>>,
    },
}
}
