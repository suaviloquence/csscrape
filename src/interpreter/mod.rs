use std::{borrow::Cow, cell::OnceCell, collections::BTreeMap, sync::Arc};

use anyhow::Context;

use crate::frontend::{
    ast::{AstRef, Element, FilterList, Leaf, RValue, SelectorOpts, Statement, StatementList},
    AstArena,
};

mod filter;
mod value;

pub use filter::Filter;
pub use value::{DataValue, TryFromValue, Value};

/// Whether we are matching a list, singular item, or optional item
/// as specified by the user
#[derive(Debug)]
enum ExecutionMode<T> {
    One(T),
    Optional(Option<T>),
    Collection(Vec<T>),
}

impl<T> ExecutionMode<T> {
    fn try_map<F, U, E>(self, mut f: F) -> Result<ExecutionMode<U>, E>
    where
        F: FnMut(T) -> Result<U, E>,
    {
        use ExecutionMode::{Collection, One, Optional};

        Ok(match self {
            One(t) => One(f(t)?),
            Optional(Some(t)) => Optional(Some(f(t)?)),
            Optional(None) => Optional(None),
            Collection(vec) => Collection(vec.into_iter().map(f).collect::<Result<_, _>>()?),
        })
    }
}

impl<'ast> Element<'ast> {
    #[must_use]
    pub fn to_selector_str(&self, ast: &AstArena<'ast>) -> String {
        use std::fmt::Write as _;

        let mut buf = String::new();
        let _ = write!(&mut buf, "{}", self.selector_head);

        for selector in ast.flatten(self.selectors) {
            let _ = write!(&mut buf, "{}", selector.sel);
        }

        buf
    }
}

#[derive(Debug, Default)]
pub struct Variables<'a, 'b>(pub BTreeMap<Cow<'a, str>, Value<'b>>);

#[derive(Debug, Default)]
pub struct DataVariables<'a>(pub BTreeMap<Cow<'a, str>, DataValue>);

impl<'a, 'b> From<Variables<'a, 'b>> for DataVariables<'a> {
    fn from(value: Variables<'a, 'b>) -> Self {
        Self(
            value
                .0
                .into_iter()
                .filter_map(|(k, v)| DataValue::try_from(v).ok().map(|v| (k, v)))
                .collect(),
        )
    }
}

impl<'a, 'b> From<DataVariables<'a>> for Variables<'a, 'b> {
    fn from(value: DataVariables<'a>) -> Self {
        Self(value.0.into_iter().map(|(k, v)| (k, v.into())).collect())
    }
}

#[derive(Debug)]
pub struct ElementContext<'ast, 'ctx> {
    variables: Variables<'ast, 'ctx>,
    element: scraper::ElementRef<'ctx>,
    text: OnceCell<Arc<str>>,
    parent: Option<&'ctx ElementContext<'ast, 'ctx>>,
}

pub fn interpret<'ast>(
    html: scraper::Html,
    ast: &'ast AstArena<'ast>,
    head: Option<AstRef<'ast, StatementList<'ast>>>,
) -> anyhow::Result<DataVariables<'ast>> {
    let mut ctx = ElementContext {
        element: html.root_element(),
        variables: Variables::default(),
        text: OnceCell::new(),
        parent: None,
    };

    for statement in ast.flatten(head) {
        interpret_statement(ast, &statement.value, &mut ctx)?;
    }

    Ok(ctx.variables.into())
}

fn interpret_statement<'ast, 'stmt, 'ctx: 'stmt>(
    ast: &'ast AstArena<'ast>,
    statement: &Statement<'ast>,
    ctx: &'stmt mut ElementContext<'ast, 'ctx>,
) -> anyhow::Result<()> {
    match statement.id {
        immutable @ ("document" | "text") => {
            anyhow::bail!("can't assign to immutable variable `{immutable}`")
        }
        id => {
            let value = match &statement.value {
                RValue::Leaf(l) => ctx.leaf_to_value(l)?,
                RValue::Element(e) => interpret_element(ctx, ctx.element, ast, e)?.into(),
            };

            let value = apply_filters(ast.flatten(statement.filters).into_iter(), value, ctx, ast)?;
            ctx.variables.0.insert(Cow::Borrowed(id), value);
        }
    }

    Ok(())
}

fn interpret_element<'ast, 'ctx>(
    parent: &'ctx ElementContext<'ast, 'ctx>,
    root: scraper::ElementRef<'ctx>,
    ast: &'ast AstArena<'ast>,
    element: &Element<'ast>,
) -> anyhow::Result<DataValue> {
    let selector_str = element.to_selector_str(ast);

    let selector = scraper::Selector::parse(&selector_str).map_err(|e| {
        anyhow::anyhow!(
            "Selector parse failed: {e}.  This is a program error. Selector is `{selector_str}`",
        )
    })?;

    let mut selection = root.select(&selector);

    let element_refs = match element.ops {
        // TODO: take the first, or fail if there are > 1?
        SelectorOpts::One => ExecutionMode::One(
            selection
                .next()
                .with_context(|| format!("Expected exactly one `{selector_str}`"))?,
        ),
        SelectorOpts::Optional => ExecutionMode::Optional(selection.next()),
        SelectorOpts::Collection => ExecutionMode::Collection(selection.collect()),
    };

    let values = element_refs.try_map(|element_ref| {
        let mut ctx = ElementContext {
            element: element_ref,
            variables: Variables::default(),
            text: OnceCell::new(),
            parent: Some(parent),
        };

        for statement in ast.flatten(element.statements) {
            let statement = &statement.value;
            interpret_statement(ast, statement, &mut ctx)?;
        }

        Ok::<_, anyhow::Error>(DataValue::Structure(
            ctx.variables
                .0
                .into_iter()
                .filter_map(|(k, v)| DataValue::try_from(v).ok().map(|v| (Arc::from(&*k), v)))
                .collect(),
        ))
    })?;

    Ok(match values {
        ExecutionMode::One(x) | ExecutionMode::Optional(Some(x)) => x,
        ExecutionMode::Optional(None) => DataValue::Null,
        ExecutionMode::Collection(l) => DataValue::List(l),
    })
}

impl<'ast, 'ctx> ElementContext<'ast, 'ctx> {
    pub fn get_var(&self, id: &str) -> anyhow::Result<Value<'ctx>> {
        match id {
            "element" => Ok(Value::Element(self.element)),
            "text" => Ok(Value::String(Arc::clone(self.text.get_or_init(|| {
                self.element
                    .children()
                    .filter_map(|x| x.value().as_text().map(|x| &**x))
                    .collect::<String>()
                    .into()
            })))),
            var => match self.variables.0.get(var) {
                Some(var) => Ok(var.clone()),
                None => self
                    .parent
                    .with_context(|| format!("Unknown variable `{var}`"))?
                    .get_var(id),
            },
        }
    }

    pub fn set_var(&mut self, name: Cow<'ast, str>, value: Value<'ctx>) -> anyhow::Result<()> {
        match &*name {
            immutable @ ("element" | "text") => {
                anyhow::bail!("Can't assign to immutable variable `{immutable}`")
            }
            _ => self.variables.0.insert(name, value),
        };

        Ok(())
    }

    pub fn leaf_to_value(&self, value: &Leaf<'ast>) -> anyhow::Result<Value<'ctx>> {
        match value {
            Leaf::Float(x) => Ok(Value::Float(*x)),
            Leaf::Int(n) => Ok(Value::Int(*n)),
            Leaf::String(s) => Ok(Value::String(Arc::from(&**s))),
            Leaf::Var(id) => self.get_var(id),
        }
    }
}

fn apply_filters<'ast, 'ctx>(
    filters: impl Iterator<Item = &'ast FilterList<'ast>>,
    value: Value<'ctx>,
    ctx: &mut ElementContext<'ast, 'ctx>,
    ast: &AstArena<'ast>,
) -> anyhow::Result<Value<'ctx>> {
    let mut value = value;
    for f in filters {
        let args: BTreeMap<_, _> = ast
            .flatten(f.args)
            .into_iter()
            .map(|a| {
                let value = match &a.value {
                    Leaf::Var(id) => ctx.get_var(id)?,
                    Leaf::Int(n) => Value::Int(*n),
                    Leaf::String(c) => Value::String(Arc::from(&**c)),
                    Leaf::Float(f) => Value::Float(*f),
                };

                Ok((a.id, value))
            })
            .collect::<anyhow::Result<_>>()?;

        value = filter::dispatch_filter(f.id, value, args, ctx)?;
    }
    Ok(value)
}

#[cfg(test)]
pub fn interpret_string_harness(
    program: &'static str,
    html: &'static str,
) -> anyhow::Result<DataVariables<'static>> {
    let (ast, head) = crate::frontend::Parser::new(program).parse()?;
    let html = scraper::Html::parse_document(html);
    interpret(html, Box::leak(Box::new(ast)), head)
}

#[cfg(test)]
mod tests {
    use super::{interpret, DataValue::*};

    fn integration_test(filename: &str) -> anyhow::Result<()> {
        let input = std::fs::read_to_string(format!("examples/inputs/{filename}.html"))?;
        let script = std::fs::read_to_string(format!("examples/scrps/{filename}.scrp"))?;
        let output: serde_json::Value = serde_json::from_reader(std::fs::File::open(format!(
            "examples/outputs/{filename}.json"
        ))?)?;

        let (ast, head) = crate::frontend::Parser::new(&script)
            .parse()
            .expect("parse error");
        let html = scraper::Html::parse_document(&input);

        let result = interpret(html, &ast, head)?;
        let result = serde_json::to_value(result.0)?;
        assert_eq!(output, result);

        Ok(())
    }

    macro_rules! integration_test {
        {
            $($name: ident,)*
        } => {
            $(
                #[test]
                fn $name() -> anyhow::Result<()> {
                    integration_test(stringify!($name))
                }
            )*
        };
    }

    #[test]
    fn test_basic() {
        let output = super::interpret_string_harness(
            r#"
            h3: h3 {
                txt: $text;

                a: a[] {
                    child: $text;
                    parent: $txt;
                };
                div: div? {};
            };
            "#,
            r#"
                <html>
                    <h3>Hello,<a>Hello, child</a><span></span>parent!</h3>
                </html>
            "#,
        )
        .expect("parsing and interpreting should succeed");

        let Some(Structure(d)) = output.0.get("h3") else {
            panic!("got {output:?}, expected h3: {{ .. }}")
        };

        let Some(List(a)) = d.get("a") else {
            panic!("got {output:?}, expected a: [ .. ]");
        };

        let Some(Structure(a)) = a.get(0) else {
            panic!("got {output:?}");
        };

        assert!(
            match a.get("parent") {
                Some(String(x)) => &**x == "Hello,parent!",
                _ => false,
            },
            "got {output:?}"
        );

        assert!(
            match a.get("child") {
                Some(String(x)) => &**x == "Hello, child",
                _ => false,
            },
            "got {output:?}"
        );

        assert!(
            match d.get("txt") {
                Some(String(x)) => &**x == "Hello,parent!",
                _ => false,
            },
            "got {output:?}"
        );
    }

    integration_test! {
        abc,
        attr,
    }
}
