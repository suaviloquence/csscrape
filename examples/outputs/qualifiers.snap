---
source: src/interpreter/mod.rs
expression: result
---
{
  "items": [
    {
      "special": null,
      "txt": "Hello! 1"
    },
    {
      "special": "special!",
      "txt": "Hello! 2"
    },
    {
      "special": null,
      "txt": "Hello! 3"
    },
    {
      "special": null,
      "txt": "Hello! 4"
    },
    {
      "special": null,
      "txt": "Hello! 5"
    },
    {
      "special": null,
      "txt": "Hello! 6"
    },
    {
      "special": null,
      "txt": "Hello! 7"
    }
  ],
  "sixth": {
    "special": null,
    "txt": "Hello! 6"
  },
  "specials": [
    null,
    "special!",
    null,
    null,
    null,
    null,
    null
  ]
}
